src/python.c

```diff
@@ -976,6 +977,124 @@ getDeparsedSortGroup(PyObject *sortKey)
 }
 
 
+/*
+ * get_environment_dict:
+ *
+ * Queries the "environment" table (which has two TEXT columns: key and value)
+ * and returns a new Python dictionary mapping keys to values.
+ *
+ * Returns: A new reference to a Python dict on success, or NULL on any failure.
+ */
+static PyObject *
+get_environment_dict(void)
+{
+    PyObject   *env_dict = NULL;
+    int         ret;
+
+    /* Create a new empty Python dictionary */
+    env_dict = PyDict_New();
+    if (env_dict == NULL) {
+        elog(WARNING, "Couldn't allocate environment data structure");
+        return NULL;
+    }
+
+    /* Connect to SPI */
+    if ((ret = SPI_connect()) != SPI_OK_CONNECT)
+    {
+        elog(WARNING, "SPI_connect failed");
+        Py_DECREF(env_dict);
+        return NULL;
+    }
+
+    /* Execute the query. 0 means no limit on the number of rows returned. */
+    /* Use PG_TRY/PG_CATCH to catch errors from SPI_exec */
+    PG_TRY();
+    {
+        ret = SPI_exec("SELECT key, value FROM environment", 0);
+    }
+    PG_CATCH();
+    {
+        ErrorData *edata = CopyErrorData();
+        /* Check for undefined table error */
+        if (edata->sqlerrcode == ERRCODE_UNDEFINED_TABLE)
+        {
+            /* Table doesn't exist; Ignore (for backward compatibility). */
+            FlushErrorState();
+            elog(DEBUG3, "No environment.");
+            SPI_finish();
+            Py_DECREF(env_dict);
+            return NULL;
```

Comment:
Is this correct? Or should it return an empty env instead? Is NULL propagate correctly across the rest of the chain? An empty environment dict seems safer to me.

Reply:
I'm trying a change to that section, making it `return env_dict` (an empty dict). And `NULL` is returned upon errors (logged).

The caller handles errors (when it gets `NULL`) by passing `None` to Python.

When there's no table:
* Log `No environment` shows in `DEBUG`,
* Python gets an empty dict.
* DAS gets `env {}` (an empty `env`).

When there's a table, but it's empty:
* No log (env was found),
* Python gets an empty dict.
* DAS gets `env {}`.

When there's an error:
* e.g. table `environment` exists but has a column of wrong type
  * `WARNING:  environment value: wrong type 23`
  * Python gets `None`
  * DAS gets no env.
* e.g. table `environment` has other columns:
  * WARNING:  Environment query failed: column "key" does not exist
  * DAS gets no env.